src/_pytest/assertion/rewrite.py

```diff
@@ -833,26 +849,47 @@ def visit_Assert(self, assert_):
                     top_condition, module_path=self.module_path, lineno=assert_.lineno
                 )
             )
-        # Create failure message.
-        body = self.on_failure
         negation = ast.UnaryOp(ast.Not(), top_condition)
-        self.statements.append(ast.If(negation, body, []))
+        msg = self.pop_format_context(ast.Str(explanation))
         if assert_.msg:
             assertmsg = self.helper("_format_assertmsg", assert_.msg)
-            explanation = "\n>assert " + explanation
+            gluestr = "\n>assert "
         else:
             assertmsg = ast.Str("")
-            explanation = "assert " + explanation
-        template = ast.BinOp(assertmsg, ast.Add(), ast.Str(explanation))
-        msg = self.pop_format_context(template)
-        fmt = self.helper("_format_explanation", msg)
+            gluestr = "assert "
+        err_explanation = ast.BinOp(ast.Str(gluestr), ast.Add(), msg)
+        err_msg = ast.BinOp(assertmsg, ast.Add(), err_explanation)
         err_name = ast.Name("AssertionError", ast.Load())
+        fmt = self.helper("_format_explanation", err_msg)
+        fmt_pass = self.helper("_format_explanation", msg)
         exc = ast_Call(err_name, [fmt], [])
         if sys.version_info[0] >= 3:
             raise_ = ast.Raise(exc, None)
         else:
             raise_ = ast.Raise(exc, None, None)
-        body.append(raise_)
+        # if self._assertion_pass:
```

Comment:
Can we even skip meddling with the AST if there are no hookimpls implementing `pytest_assertion_pass`? I fear bugs on the assertion rewriting code will sprung once this is used in the wild and we might need to revert this feature in a rush like we did recently with the `all()` unrolling.

Reply:
Rewriting conditional on hookimpls were my original idea but it might cause problem with caching, as discussed in the beginning: Users would need to clear the pyc cache everytime it adds a plugin that implements it or uses it himself in a `conftest.py`.

This might be a big problem or not, what do you think?

Reply:
We could do a assertion rewriting switch like this but based on a command line option, which could be disabled by default while this hook is experimental (we can document that and even detect the hookimpls that this will only work if they enable via command line)

Reply:
> Rewriting conditional on hookimpls were my original idea but it might cause problem with caching, as discussed in the beginning: Users would need to clear the pyc cache everytime it adds a plugin that implements it or uses it himself in a conftest.py.

I think this will have few uses initially, so actually I'm fine with this option; we can document it very prominently in the docs that this is necessary on activation.

To reiterate, I think this feature might spring some cool plugins, but I fear that this might affect users that don't use the hook at all, and assertion rewriting is pretty tricky as you know. 😁